include/SampleCache.h

```diff
@@ -0,0 +1,125 @@
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_SAMPLE_CACHE_H
+#define LMMS_SAMPLE_CACHE_H
+
+#include <QString>
+#include <filesystem>
+#include <memory>
+#include <unordered_map>
+
+#include "SampleBuffer.h"
+
+namespace lmms {
+class SampleCache
+{
+public:
+	/**
+		Fetches a sample from the cache through a path to an audio file,
+		and returns the stored buffer.
+
+		If `path` exists in the cache, its last write time is checked with what is currently in the cache. If
+		there is a mismatch, the sample is reloaded from disk, its entry in the cache is updated, and the sample is
+		returned.
+
+		If `path` does not exist in the cache, the sample is loaded from disk and
+		then returned.
+	 */
+	static auto fetch(const QString& path) -> std::shared_ptr<SampleBuffer>;
+
+	/**
+		Fetches a sample from the cache through a Base64 string and a sample rate
+		and returns the stored buffer.
+
+		If an entry for a `base64` string with a certain `sampleRate` exists in the cache, the stored sample is
+		returned. Otherwise, if it does not exist in the cache, the sample is loaded and then returned.
+	 */
+	static auto fetch(const QString& base64, int sampleRate) -> std::shared_ptr<SampleBuffer>;
+
+private:
+	struct AudioFileEntry
+	{
+		friend bool operator==(const AudioFileEntry& first, const AudioFileEntry& second) noexcept
+		{
+			return first.path == second.path && first.lastWriteTime == second.lastWriteTime;
+		}
+
+		std::filesystem::path path;
+		std::filesystem::file_time_type lastWriteTime;
+	};
+
+	struct Base64Entry
+	{
+		friend bool operator==(const Base64Entry& first, const Base64Entry& second) noexcept
+		{
+			return first.base64 == second.base64 && first.sampleRate == second.sampleRate;
+		}
+
+		std::string base64;
+		int sampleRate;
+	};
+
+	struct Hash
+	{
+		std::size_t operator()(const AudioFileEntry& entry) const noexcept
+		{
+			return std::filesystem::hash_value(entry.path);
+		}
+
+		std::size_t operator()(const Base64Entry& entry) const noexcept
+		{
+			return std::hash<std::string>()(entry.base64);
+		}
+	};
+
+	template <typename T, typename ...Args>
+	static auto get(const T& entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args... args)
+	{
+		const auto it = map.find(entry);
+
+		if (it == map.end())
+		{
+			const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
+			map.insert(std::make_pair(entry, buffer));
+			return buffer;
+		}
+
+		const auto entryLock = it->second.lock();
+		if (!entryLock)
+		{
+			const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
+			map[entry] = buffer;
+			return buffer;
+		}
+
+		return entryLock;
+	}
```

Comment:
This function makes a copy of `entry` when it could be moved into this function from the call sites to avoid copying the file path or base64 data.

If the first parameter was `T entry`, you could emplace into the maps
```cpp
map.emplace(std::move(entry), buffer);
```